/*
* Help message
*/

def helpMessage() {
    println ''
    log.info """
    nucMACC   P I P E L I N E
    =============================
    Usage:

    nextflow run uschwartz/nucMACC --analysis 'typeOfAnalysis' --csvInput 'path2csvFile' --outDir 'path2outDir' --genomeIdx 'path2bowtie2_idx' --genomeSize 'eff. genome size' --genome 'path2ref_genome_fasta'

    General options:
      --analysis            [string] type of analysis (default: nucMACC; possible inputs: "nucMACC" or "MNaseQC")
      --test                [boolean]if set, test run will be perfomed with toyData instead; csvInput for "bamEntry" needs to be created prior! (default: false)
      --genomeSize          [integer] Effective genome size, defined as the length of the mappable genome. Used for normalisation (default: 162367812 (dm3))

    Path variables used in analysis:
      --csvInput            [string] Path to comma-separated file containing information about the samples in the experiment (see ./toyData/input.csv as example) template provided in ./input_template.csv
      --outDir              [string] Name of output directory, which will be created (default: ~/nucMACC_test/)
      --genomeIdx           [string] Path and prefix of bowtie2 index (minus .X.bt2)
      --genome              [string] Path to reference genome in fasta Format

    Additional options:
      --blacklist           [string] A BED file containing regions that should be excluded from all nucleosome analysis (default: false)
      --TSS                 [string] A Transcript annotation file in GTF format to obtain TSS profiles. For more details about TSS format see documentation of deeptools computeMatrix, which is used here (default: false)
      --publishBam          [boolean]if set, aligned bam files will be written to outDir (default: false)
      --publishBamFlt       [boolean]if set, size selected bam files will be written to outDir (default: false)

    nucMACC specific options:
      --bamEntry            [boolean] if set, bamEntry point for nucMACC using .bam-files generated by prior MNaseQC run will be used, instead of default FASTQ entry (default: false)

    MNaseQC specific options:
      --correlationMethod   [string] choose correlation method for correlation heatmap in MNase_QC (default: "spearman", possible inputs: "spearman" or "pearson")

     """.stripIndent()
     println ''
}
